kernel: cbs: adding support for the Constant Bandwidth Server (CBS)

alexpaschoaletto opened the pull request:
This PR introduces kernel support for the Constant Bandwidth Server (CBS).

A CBS is an extension of the Earliest Deadline First (EDF) scheduler already implemented (although with some limitations) in [Linux](https://docs.kernel.org/scheduler/sched-deadline.html) and [RTEMS](https://docs.rtems.org/branches/master/c-user/constant_bandwidth_server.html) that allows tasks to be executed virtually isolated from each other (time-wise, not memory-wise), in a way that if a task executes for longer than expected it doesn't interfere on the execution of the others. In other words, the CBS prevents that a task
misbehavior causes other tasks to miss their own deadlines.

In a nutshell, the CBS is a work-conserving wrapper for the tasks that automatically recalculates their deadlines when they exceed their configured timing constraints. If a task takes longer than expected, it can be preempted by awaiting tasks instead of jeopardizing their deadlines.

You want to use the CBS when:

1. You want an automated way of recalculating the task deadlines.
2. You have a task set running under EDF and there needs to be a *scheduling guarantee*, that is, a guarantee that no task will ever starve because a higher priority one is executing for too long.
3. You want to improve the response time of lower priority tasks without jeopardizing the guarantees of the higher priority tasks.
4. You want to decide the task priorities by defining their guaranteed CPU time (for example, task 1 needs 50%, task 2 needs 30% and task 3 needs 20%), like a round-robin on steroids. That is possible because the ratio of the CBS attributes (budget, period) is equivalent to the guaranteed bandwidth of the server. It should be noted that if that's the case, the sum of all tasks' utilization should necessarily be less or equal than 100%.

So, depending on what you need, the CBS can be seen as a safe way of serving tasks with no deadlines within a EDF taskset (like a workqueue, but better) **or** a fail-safe for existing EDF tasks that might overrun, and thus endanger other tasks's deadlines.

The most important implementation file to analyze is [this one](https://github.com/alexpaschoaletto/zephyr/blob/cbs/kernel/sched_server/cbs.c). The rest are docs, samples, and eventual/small kernel mods to enable the CBS to work accordingly.

For more details, feel free to read RFC #78000 and the [docs](https://github.com/alexpaschoaletto/zephyr/blob/cbs/doc/kernel/services/sched_server/cbs.rst) wrote for this contribution. Or just ask, I'll be happy to provide an answer.

Commit:
kernel: CBS: Adding support for Constant Bandwidth Server (CBS)

This commit introduces kernel support for the Constant
Bandwidth Server (CBS), an extension of the Earliest
Deadline First (EDF) scheduler that allows tasks to be
executed virtually isolated from each other, in a way
that if a task executes for longer than expected it doesn’t
interfere on the execution of the others. In other words,
the CBS prevents that a task misbehavior causes other tasks
to miss their own deadlines.

In a nutshell, the CBS is a work-conserving wrapper for
the tasks that automatically recalculates their deadlines
when they exceed their configured timing constraints, thus
acting as a fail-safe for EDF overruns.

Signed-off-by: Alexander pinheiro paschoaletto <1222703@isep.ipp.pt>

andyross commented on subsys/sched_server/cbs/cbs.c:
```diff
@@ -0,0 +1,333 @@
+		return;
+	}
+	cbs_cycle_t now = cbs_get_now();
+	cbs_t *cbs = (cbs_t *)k_timer_user_data_get(timer);
```
FWIW if you make the k_timer a cbs field, you can just use a CONTAINER_OF() here to get the timer pointer, which is a single addition.

alexpaschoaletto replied:
great suggestion! done.

andyross commented on include/zephyr/sched_server/cbs.h:
```diff
@@ -0,0 +1,162 @@
+typedef struct {
+	cbs_callback_t function;
+	void *arg;
+} cbs_job_t;
```
No typedef struct names in zephyr please, absent special circumstances like by-value opaque types.  Should just be "struct cbs_job"

alexpaschoaletto replied:
ok, I've removed typedef structs.

andyross commented on subsys/sched_server/cbs/cbs.c:
```diff
@@ -0,0 +1,333 @@
+
+void cbs_thread_switched_in(struct k_thread *thread)
+{
+	if (!thread || !thread->cbs) {
```
Why check thread for null here?  That would be a bug in the context switch primitives you hooked.  This is **extremely** performance-sensitive code.

alexpaschoaletto replied:
Sorry, it's just me used to include defensive code by default. I have removed that.

andyross commented on subsys/sched_server/cbs/cbs.c:
```diff
@@ -0,0 +1,333 @@
+		return;
+	}
+	cbs_t *cbs = (cbs_t *)thread->cbs;
+	if (cbs->is_initialized && !cbs->is_idle) {
```
Likewise no defensive coding in the context switch path please.  If you want to be sure cbs is initialized then the architecture needs to be affirmatively sure it is before that pointer is set on the thread.  It's OK to make this an assert if you're worried about design bugs.

alexpaschoaletto replied:
It's ok, this code wasn't needed indeed. I have removed it.

andyross reviewed (commented):
Some quick notes after a quick and incomplete read.  The big worry here is performance: this is adding two function calls and a bunch of tests to the context switch path for one somewhat obscure feature, and that's sort of a non-starter.  At the very least, this should emit code that does an absolute minimum of work (e.g. just check a boolean on the thread struct) before returning into the main switch path.  Calling out into a separate stack frame is spending a ton of cycles we can't spare.

Does the problem really require tracking budget across cooperative switches anyway?   Can't you do this with a timer a-la timeslicing?  Seems like this is taking a ton of overhead for what would normally seem like a fairly coarse decision.  Thread deadlines are usually on the order of a tick, not a cycle.

Commit:
samples: CBS: adding code samples for the Constant Bandwidth Server (CBS)

This commit adds two sample codes for the the CBS,
one for testing a very simple logging application
and other for effectively implementing it alongside
other hard real-time threads.

Signed-off-by: Alexander pinheiro paschoaletto <1222703@isep.ipp.pt>

alexpaschoaletto commented on kernel/sched.c:
```diff
@@ -835,8 +839,16 @@ struct k_thread *z_swap_next_thread(void)
 static inline void set_current(struct k_thread *new_thread)
 {
+	#ifdef CONFIG_CBS
+	cbs_thread_switched_out(_current_cpu->current);
```
Can you confirm this information? I'm applying the fixes but the latest `main` branch seems to have it no more. It looks like `_current` is being used [instead](https://github.com/zephyrproject-rtos/zephyr/blob/d4deeebff1870c3aa7372c14d403120f8a619282/include/zephyr/kernel_structs.h#L267).

nashif commented:
quick note here, you should not be introducing a core scheduling/kernel feature as a new subsystem, this belongs into the already existing kernel "subsystem"

alexpaschoaletto commented:
> quick note here, you should not be introducing a core scheduling/kernel feature as a new subsystem, this belongs into the already existing kernel "subsystem"

I can move into the main kernel files if that's no problem.

Commit:
docs: CBS: adding docs for the Constant Bandwidth Server (CBS)

This commit simply adds the documentation for the CBS.

Signed-off-by: Alexander pinheiro paschoaletto <1222703@isep.ipp.pt>

alexpaschoaletto commented on include/zephyr/sched_server/cbs.h:
```diff
@@ -0,0 +1,162 @@
+} cbs_arg_t;
+
+typedef struct {
+	struct k_timer *timer;
```
done!

alexpaschoaletto commented on kernel/thread.c:
```diff
@@ -942,6 +945,7 @@ void z_thread_mark_switched_out(void)
 }
 #endif /* CONFIG_INSTRUMENT_THREAD_SWITCHING */
 
+
```
sorry 🫠 removed.

alexpaschoaletto commented on subsys/sched_server/cbs/cbs.c:
```diff
@@ -0,0 +1,333 @@
+	cbs_arg_t *args = (cbs_arg_t *)cbs_args;
+
+	cbs->thread = k_current_get();
+	cbs->thread->cbs = cbs;
```
well, `cbs->thread` is needed for when if the CBS thread demands a deadline updated (when a job is pushed to it, see function `k_cbs_push_job` and `cbs_check_current_bandwidth` it calls). I can't think of a better way of fetching the CBS thread if not like this.

The `thread->cbs` is needed for the context switches; the cbs timer should start when the CBS thread switches in and stop when it switches out. And again, I can't think of a better way of fetching the CBS thread if not like this.

I'm open to suggestions if you wish.

alexpaschoaletto commented on subsys/sched_server/cbs/cbs.c:
```diff
@@ -0,0 +1,333 @@
+	 * then the configured bandwidth. If that
+	 * is true, the job can be served with the
+	 * current budget and deadline. If not,
+	 * the deadline needs to be postponed.
```
Alright, I have refactored the comments to increase line lengths.

alexpaschoaletto commented on subsys/sched_server/cbs/cbs.c:
```diff
@@ -0,0 +1,333 @@
+	}
+	cbs_cycle_t now = cbs_get_now();
+	cbs_t *cbs = (cbs_t *)k_timer_user_data_get(timer);
+	k_sched_lock();
```
ok-dokey, I have changed to `K_SPINLOCK`.

alexpaschoaletto commented:
I'd like the label removed, please.

cfriedt commented on kernel/Kconfig.sched_server:
```diff
@@ -0,0 +1,130 @@
+			Scheduling delay (in milliseconds) for the CBS thread.
+
+	if CBS_LOG
+	config CBS_THREAD_MAX_NAME_LEN
```
This likely duplicates something that already exists in Zephyr, so I would suggest using the existing convention.

cfriedt commented on kernel/Kconfig.sched_server:
```diff
@@ -0,0 +1,130 @@
+		requires CONFIG_CBS to be included within the project.
+
+if CBS_LOG
+menu "CBS events to be logged"
```
The events being recorded here really seem like they would fit better with Zephyr's tracing subsystem. Can they use that mechanism instead?

cfriedt reviewed (commented):
Just a quick fly-by. I can review the actual scheduler code soon.

alexpaschoaletto commented:
> Just a quick fly-by. I can the actual scheduler code soon.

Thank you so much for taking your time, cfriedt! I'm currently busy with other stuff at the moment, but I will get back to this eventually and address all your comments.
